### PR TITLE
11314 create joint properties widget

### PR DESCRIPTION
**Pull request by ablepharus**
## What does this PR do?

Resolves https://github.com/o3de/o3de/issues/11314

image

## How was this PR tested?

manually

### CODE REVIEW

File: Gems/EMotionFX/Code/Source/Editor/Plugins/SkeletonOutliner/JointPropertyWidget.cpp [hunk 1]
```diff
@@ -0,0 +1,105 @@
+#include "JointPropertyWidget.h"
```
Comment (cgalvan) [inline Gems/EMotionFX/Code/Source/Editor/Plugins/SkeletonOutliner/JointPropertyWidget.cpp:1]: A few of these files are missing the copyright headers at the top, which has failed vaildation:

```
VALIDATOR_FAILED: CopyrightHeaderValidator /home/runner/work/o3de/o3de/Gems/EMotionFX/Code/Source/Editor/Plugins/SkeletonOutliner/JointPropertyWidget.cpp::CopyrightHeaderValidator FAILED - Source file missing O3DE copyright header.
VALIDATOR_FAILED: CopyrightHeaderValidator /home/runner/work/o3de/o3de/Gems/EMotionFX/Code/Source/Editor/Plugins/SkeletonOutliner/JointPropertyWidget.cpp::CopyrightHeaderValidator FAILED - Source file missing any recognized copyrights.
VALIDATOR_FAILED: CopyrightHeaderValidator /home/runner/work/o3de/o3de/Gems/EMotionFX/Code/Source/Editor/Plugins/SkeletonOutliner/JointPropertyWidget.h::CopyrightHeaderValidator FAILED - Source file missing O3DE copyright header.
VALIDATOR_FAILED: CopyrightHeaderValidator /home/runner/work/o3de/o3de/Gems/EMotionFX/Code/Source/Editor/Plugins/SkeletonOutliner/JointPropertyWidget.h::CopyrightHeaderValidator FAILED - Source file missing any recognized copyrights.
```

**Review by cgalvan — CHANGES_REQUESTED** [review]

File: Gems/EMotionFX/Code/Source/Editor/Plugins/SkeletonOutliner/JointPropertyWidget.cpp [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+#include <QItemSelectionModel>
+#include <QBoxLayout>
+#include <QLabel>
+
+namespace EMotionFX
+{
+    JointPropertyWidget::JointPropertyWidget(QWidget* parent)
+        :QWidget(parent)
+    {
+        auto mainLayout = new QVBoxLayout;
```
Comment (amzn-jillich) [inline Gems/EMotionFX/Code/Source/Editor/Plugins/SkeletonOutliner/JointPropertyWidget.cpp:20]: Nit: * missing

File: Gems/EMotionFX/Code/Source/Editor/Plugins/SkeletonOutliner/JointPropertyWidget.cpp (hunk 1 above)
Comment (amzn-jillich) [inline Gems/EMotionFX/Code/Source/Editor/Plugins/SkeletonOutliner/JointPropertyWidget.cpp:1]: Nit: Use <>

**Review by amzn-jillich — APPROVED** [review]

**Comment by amzn-jillich** [comment]
Can you please post a video of how it looks like now when selecting the character vs. selecting a joint and without any selection etc?

File: Gems/EMotionFX/Code/Source/Editor/Plugins/SkeletonOutliner/SkeletonOutlinerPlugin.h [hunk 3]
```diff
@@ -17,6 +17,7 @@
 #include <Editor/SelectionProxyModel.h>
 #include <Editor/SkeletonSortFilterProxyModel.h>
 #include <QTreeView>
+#include "JointPropertyWidget.h"
```
Comment (amzn-rhhong) [inline Gems/EMotionFX/Code/Source/Editor/Plugins/SkeletonOutliner/SkeletonOutlinerPlugin.h:20]: Use #include <xxx> style

File: Gems/EMotionFX/Code/Source/Editor/Plugins/SkeletonOutliner/JointPropertyWidget.h [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+#include <QPushButton>
+#endif
+namespace EMStudio {
+    class ActorInfo;
+    class NodeInfo;
+}
+namespace EMotionFX
+{
+    class SkeletonOutlinerPlugin;
+    class AttributesWindowRequests
```
Comment (amzn-rhhong) [inline Gems/EMotionFX/Code/Source/Editor/Plugins/SkeletonOutliner/JointPropertyWidget.h:21]: can you add a comment in here explaining what the class does

File: Gems/EMotionFX/Code/Source/Editor/Plugins/SkeletonOutliner/JointPropertyWidget.h [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        : public AZ::EBusTraits
+    {
+    public:
+        virtual bool IsLocked() const { return false; }
+        virtual QModelIndex GetModelIndex() const { return {}; }
+    };
+
+    using AttributesWindowRequestBus = AZ::EBus<AttributesWindowRequests>;
+
+    class JointPropertyWidget
```
Comment (amzn-rhhong) [inline Gems/EMotionFX/Code/Source/Editor/Plugins/SkeletonOutliner/JointPropertyWidget.h:31]: same here, add a few comments explain what the class does.

File: Gems/EMotionFX/Code/Source/Editor/Plugins/SkeletonOutliner/JointPropertyWidget.h [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+#include <Editor/Plugins/Cloth/ClothJointWidget.h>
+#include <Editor/Plugins/HitDetection/HitDetectionJointWidget.h>
+#include <UI/PropertyEditor/ReflectedPropertyEditor.hxx>
+#include <EMotionFX/Tools/EMotionStudio/EMStudioSDK/Source/DockWidgetPlugin.h>
+#include <Editor/Plugins/SkeletonOutliner/SkeletonOutlinerBus.h>
+#include <AtomToolsFramework/Inspector/InspectorWidget.h>
+#include <Editor/Plugins/Ragdoll/RagdollNodeWidget.h>
+#include <QPushButton>
+#endif
+namespace EMStudio {
```
Comment (amzn-rhhong) [inline Gems/EMotionFX/Code/Source/Editor/Plugins/SkeletonOutliner/JointPropertyWidget.h:14]: please add an empty line between line 13 and line 14, also between line 17 and 18.

File: Gems/EMotionFX/Code/Source/Editor/Plugins/SkeletonOutliner/JointPropertyWidget.cpp [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+        Node* node{nullptr};
+        SkeletonOutlinerRequestBus::BroadcastResult(node, &SkeletonOutlinerRequests::GetSingleSelectedNode);
+
+        if (node)
+        {
+            m_nodeInfo.reset(aznew EMStudio::NodeInfo(actorInstance, node));
+            m_propertyWidget->AddInstance(m_nodeInfo.get(), azrtti_typeid(m_nodeInfo.get()));
+        }
+        else if(actorInstance && actorInstance->GetActor() )
```
Comment (amzn-rhhong) [inline Gems/EMotionFX/Code/Source/Editor/Plugins/SkeletonOutliner/JointPropertyWidget.cpp:82]: extra space at getActor() )

**Review by amzn-rhhong — COMMENTED** [review]
minor comments

**Review by cgalvan — APPROVED** [review]

File: Gems/EMotionFX/Code/Source/Editor/Plugins/SkeletonOutliner/JointPropertyWidget.h (hunk 4 above)
Comment (ablepharus) [reply Gems/EMotionFX/Code/Source/Editor/Plugins/SkeletonOutliner/JointPropertyWidget.h:21]: Thats a good question. This made sense at one point, but is not needed anymore and gets scrapped.

File: Gems/EMotionFX/Code/Source/Editor/Plugins/SkeletonOutliner/JointPropertyWidget.h (hunk 5 above)
Comment (ablepharus) [reply Gems/EMotionFX/Code/Source/Editor/Plugins/SkeletonOutliner/JointPropertyWidget.h:31]: "A Widget in the Inspector Pane displaying Attributes of selected Nodes in a Skeleton"  good enough?

**Comment by hultonha** [comment]
@amzn-rhhong - we discussed this issue with @ablepharus and there is a crash happening when selecting the actor. Is this something you could help them diagnose please? Thanks!

File: Gems/EMotionFX/Code/Source/Editor/Plugins/SkeletonOutliner/JointPropertyWidget.h (hunk 5 above)
Comment (hultonha) [reply Gems/EMotionFX/Code/Source/Editor/Plugins/SkeletonOutliner/JointPropertyWidget.h:31]: That sounds good to me :+1:

**Comment by ablepharus** [comment]
[11314-create-joint-properties-widget.webm](https://user-images.githubusercontent.com/57981/192858479-cb1778dc-3cea-40e5-bc48-43bbe6ab6de5.webm)

**Comment by hultonha** [comment]
How did you get on investigating the crash @ablepharus? This PR needs to be merged before https://github.com/o3de/o3de/pull/12041. It would be great to get to the bottom of it and then get this then 12041 merged in quick succession. Thanks!

File: Gems/EMotionFX/Code/Source/Editor/Plugins/SkeletonOutliner/JointPropertyWidget.h [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    public:
+        JointPropertyWidget(QWidget* parent = nullptr);
+        ~JointPropertyWidget();
+
+    public slots:
+        void Reset();
+    private:
+        AzToolsFramework::ReflectedPropertyEditor*      m_propertyWidget;
+        AZStd::unique_ptr<EMStudio::ActorInfo>          m_actorInfo;
+        AZStd::unique_ptr<EMStudio::NodeInfo>           m_nodeInfo;
```
Comment (hultonha) [inline Gems/EMotionFX/Code/Source/Editor/Plugins/SkeletonOutliner/JointPropertyWidget.h:47]: nit: Avoid additional indentation (alignment), just keep one space

File: Gems/EMotionFX/Code/Source/Editor/Plugins/SkeletonOutliner/JointPropertyWidget.h [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    };
+
+}
```
Comment (hultonha) [inline Gems/EMotionFX/Code/Source/Editor/Plugins/SkeletonOutliner/JointPropertyWidget.h:50]: nit: Closing namespace name

**Review by hultonha — COMMENTED** [review]
A few minor comments but nothing too major, let me know if you want to do a call to discuss the crash issue too :+1:

**Comment by hultonha** [comment]
Also tagging @greerdv in case they'd be interested in any of the joint changes

File: Gems/EMotionFX/Code/Source/Editor/Plugins/SkeletonOutliner/JointPropertyWidget.cpp [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+            connect(skeletonModel, &SkeletonModel::modelReset, this, &JointPropertyWidget::Reset);
+            connect(&skeletonModel->GetSelectionModel(), &QItemSelectionModel::selectionChanged,
+                    this, &JointPropertyWidget::Reset);
+        }
+
+    }
+
+    JointPropertyWidget::~JointPropertyWidget()
+    {
+        EMStudio::InspectorRequestBus::Broadcast(&EMStudio::InspectorRequestBus::Events::ClearIfShown, this);
```
[the remark this answers is not in the document]
Comment (ablepharus) [reply Gems/EMotionFX/Code/Source/Editor/Plugins/SkeletonOutliner/JointPropertyWidget.cpp:60]: I'm confused. What do you mean?

File: Gems/EMotionFX/Code/Source/Editor/Plugins/SkeletonOutliner/JointPropertyWidget.cpp [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+
+        if (m_propertyWidget)
+        {
+            delete m_propertyWidget;
+        }
```
[the remark this answers is not in the document]
Comment (ablepharus) [reply Gems/EMotionFX/Code/Source/Editor/Plugins/SkeletonOutliner/JointPropertyWidget.cpp:65]: ok

**Comment by ablepharus** [comment]
> How did you get on investigating the crash @ablepharus? This PR needs to be merged before #12041. It would be great to get to the bottom of it and then get this then 12041 merged in quick succession. Thanks!

The Actor Node has its indices set to InvalidIndex, I checked for that and used Actor Info instead.

File: Gems/EMotionFX/Code/Source/Editor/Plugins/SkeletonOutliner/JointPropertyWidget.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (hultonha) [reply Gems/EMotionFX/Code/Source/Editor/Plugins/SkeletonOutliner/JointPropertyWidget.cpp:60]: Oh sorry this was weird formatting in the GitHub UI, please ignore sorry

**Comment by hultonha** [comment]
> > How did you get on investigating the crash @ablepharus? This PR needs to be merged before #12041. It would be great to get to the bottom of it and then get this then 12041 merged in quick succession. Thanks!
> 
> The Actor Node has its indices set to InvalidIndex, I checked for that and used Actor Info instead.

Ah I see, nice find. Thanks for the update @ablepharus 👍

**Comment by Be-ing** [comment]
This also implements https://github.com/o3de/o3de/issues/11315

**Comment by hultonha** [comment]
> > Did you checkout this branch to verify?
> 
> Yes, the screenshot I posted was taken using this branch.

Perfect thanks, a build is running now, hopefully it succeeds, thanks for the info :+1:

**Comment by hultonha** [comment]
Hi @Be-ing, it looks like the last build failed for a legit reason to do with your change - https://jenkins.build.o3de.org/blue/organizations/jenkins/O3DE/detail/PR-12060/4/pipeline. Can you please investigate and see if it can be resolved. Thank you!

**Comment by Be-ing** [comment]
I think you commented on the wrong PR? Did you mean https://github.com/o3de/o3de/pull/12429?

**Comment by hultonha** [comment]
> I think you commented on the wrong PR? Did you mean #12429? Or did you intend to ping @ablepharus?

Apologies I meant to tag @ablepharus, the failure is for this PR (12060 in the URL)

File: Gems/EMotionFX/Code/Source/Editor/Plugins/SkeletonOutliner/SkeletonOutlinerPlugin.h [hunk 12]
```diff
@@ -78,7 +80,7 @@ namespace EMotionFX
         SkeletonSortFilterProxyModel*           m_filterProxyModel;
         static constexpr int s_iconSize = 16;
 
-        JointPropertyWidget*    m_propertyWidget;
+        JointPropertyWidget* m_propertyWidget{0};
```
Comment (hultonha) [inline Gems/EMotionFX/Code/Source/Editor/Plugins/SkeletonOutliner/SkeletonOutlinerPlugin.h:83]: nit: We generally just prefer `JointPropertyWidget* m_propertyWidget = nullptr;`

**Review by cgalvan — APPROVED** [review]

File: Code/Framework/AzQtComponents/AzQtComponents/Components/Widgets/Card.cpp [hunk 13]
```diff
@@ -27,6 +27,7 @@ namespace AzQtComponents
 {
 
     static QString g_containerCardClass = QStringLiteral("ContainerCard");
+    static QString g_sectionCardClass = QStringLiteral("SectionCard");
```
Comment (pollend) [inline Code/Framework/AzQtComponents/AzQtComponents/Components/Widgets/Card.cpp:30]: do we want to make this a static const?

**Pull request merged**